### DISCUSSION TITLE
Muon+ Doesn't Improve Validation Loss

### DISCUSSION THREAD

**Discussion by jon123boss** [KellerJordan/modded-nanogpt#239] [General]
Hello wonderful people of this community! I came across this [paper](https://www.alphaxiv.org/abs/2602.21545) that came out less than a day ago claiming to improve upon NorMuon and implemented it in the current code base (see attached train_gpt.py). Using the best performing "col_row" normalisation method caused an additional 0.005 increase in loss (around the same as setting beta2 in NorMuon to 0) with negligible decrease in time. 

[train_gpt.py](https://github.com/user-attachments/files/25597794/train_gpt.py)

Specifically this was added to the NorMuonAndAdam class

```
    @staticmethod
    @torch.compile(dynamic=False, fullgraph=True)
    def _apply_muon_plus_norm(x: Tensor, norm_dir: str, eps: float = 1e-8):
        """
        MUON+ Normalization step. 
        Operates in float32 for precision. Supports col, row, col_row, row_col.
        """
        x_float = x.float()
        if norm_dir == "col":
            denom = (x_float.square().sum(dim=-2, keepdim=True) + eps).sqrt()
            return x_float / denom
        elif norm_dir == "row":
            denom = (x_float.square().sum(dim=-1, keepdim=True) + eps).sqrt()
            return x_float / denom
        elif norm_dir == "col_row":
            # Apply Col Norm then Row Norm
            denom = (x_float.square().sum(dim=-2, keepdim=True) + eps).sqrt()
            x_float = x_float / denom
            denom = (x_float.square().sum(dim=-1, keepdim=True) + eps).sqrt()
            return x_float / denom
        elif norm_dir == "row_col":
            # Apply Row Norm then Col Norm
            denom = (x_float.square().sum(dim=-1, keepdim=True) + eps).sqrt()
            x_float = x_float / denom
            denom = (x_float.square().sum(dim=-2, keepdim=True) + eps).sqrt()
            return x_float / denom
        return x_float
```
 and was called instead of the variance reduction
```
if p_cfg.norm_dir:
            # MUON+ Normalization
            v_chunk = NorMuonAndAdam._apply_muon_plus_norm(v_chunk, p_cfg.norm_dir)
        else:
            # Variance reduction
            red_dim = -1 if chunk_shape[-2] >= chunk_shape[-1] else -2
            v_chunk = NorMuonAndAdam._apply_normuon_variance_reduction(
                v_chunk, p_state["second_momentum_buffer"], p_cfg.beta2, red_dim
            )
```
What do you guys think? Is my implementation correct?

**Comment by koute** [comment]
FWIW, I just quickly tried this in my own trainer fine-tuning a 600M model on 20M tokens, and it definitely works.

- Final loss without Muon+: 3.9138
- Final loss with Muon+ (using 'row' normalization): 3.7909
- Final loss with Muon+ (using 'col' normalization): 3.8096
- Final loss with Muon+ (using 'col-row' normalization): 3.8122
- Final loss with Muon+ (using 'row-col' normalization): 3.8325

Caveat: I don't use NorMuon in the baseline, this is a real-world fine-tuning run (as opposed to training from scratch), and my trainer isn't benchmaxxed towards a specific single task like modded-nanogpt is.

**Reply by jon123boss** [reply]
Thank you for your response! It is very interesting that "row" normalization resulted in the lowest final loss and it doesn't seem to be close. Do you have an explanation for this? Is it possible for you to try using NorMuon as the baseline?

Currently my best guess to why Muon+ underperformed when compared to the results in the paper is that NorMuon preserves some signal about the gradient magnitude via the second moment. If different layers in the model require different update scales (due to different learning rates), NorMuon handles this automatically, while Muon+ cannot.

**Reply by koute** [reply]
> Do you have an explanation for this?

No idea. Might be because I'm fine-tuning an existing model instead of training from scratch (i.e. the weights are not random so row-wise vs column-wise normalization has more of an impact).

> Is it possible for you to try using NorMuon as the baseline?

Here's the loss with NorMuon (code copy-pasted from the reference implementation, beta2 is 0.95): 3.9439

So even worse than vanilla for some reason?

**Reply by jon123boss** [reply]
Thanks for running that baseline so quickly! That NorMuon result being worse than vanilla is definitely unexpected and definitely could be due to being in the fine-tuning regime.
